subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ErrorHandlingModuleComponentRepositoryTest.groovy

```diff
@@ -0,0 +1,237 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.artifacts.ivyservice.ivyresolve
+
+import org.gradle.api.artifacts.component.ComponentArtifactIdentifier
+import org.gradle.api.artifacts.component.ComponentIdentifier
+import org.gradle.api.artifacts.component.ModuleComponentIdentifier
+import org.gradle.api.internal.component.ArtifactType
+import org.gradle.internal.component.external.model.DefaultModuleComponentSelector
+import org.gradle.internal.component.model.ComponentArtifactMetadata
+import org.gradle.internal.component.model.ComponentOverrideMetadata
+import org.gradle.internal.component.model.ComponentResolveMetadata
+import org.gradle.internal.component.model.DependencyMetadata
+import org.gradle.internal.component.model.ModuleSource
+import org.gradle.internal.resolve.ArtifactResolveException
+import org.gradle.internal.resolve.ModuleVersionResolveException
+import org.gradle.internal.resolve.result.BuildableArtifactResolveResult
+import org.gradle.internal.resolve.result.BuildableArtifactSetResolveResult
+import org.gradle.internal.resolve.result.BuildableComponentArtifactsResolveResult
+import org.gradle.internal.resolve.result.BuildableModuleComponentMetaDataResolveResult
+import org.gradle.internal.resolve.result.BuildableModuleVersionListingResolveResult
+import spock.lang.Specification
+import spock.lang.Subject
+
+class ErrorHandlingModuleComponentRepositoryTest extends Specification {
```

Comment:
❌ This test is over-asserting so much that it basically repeats the implementation line by line. It also tests many interactions in every method. This will break on all kinds of internal refactorings that don't change behavior at all.

A unit test should only test one interaction at a time. For everything that the test method does not care about, a stub should be used instead of a mock.

E.g. there should be one test asserting that the repository access blacklists a repo when it threw an exception. That test should assert nothing else.

For simple types like our BlackLister it is probably easier to use state-based testing, since we already have the appropriate query methods. I.e. use the real implementation and its query method instead of a mock.

Reply:
I've broken it into small piece code of `when-then` blocks here: https://github.com/gradle/gradle/blob/79dd5b101a208061bfd43d8620d81fc70d94027d/subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/ErrorHandlingModuleComponentRepositoryTest.groovy

Hope this would be better.